### PR TITLE
[PR](Network): implement ByteConverter and CircularBuffer classes with serialization and deserialization methods

### PR DESCRIPTION
**Pull request by ctalneau**
This pull request introduces two new utility classes to the networking module: `ByteConverter` for serialization and deserialization of basic types, and `CircularBuffer` for efficient byte storage and retrieval. These additions are integrated into the codebase and thoroughly tested with new unit tests.

### Serialization and Buffer Utilities

* Added the `ByteConverter` class to `include/rtype/network/ByteConverter.hpp` and implemented its methods in `src/network/ByteConverter.cpp` for serializing and deserializing `int32_t`, `float`, and `std::string` types to and from byte buffers. [[1]](diffhunk://#diff-133171faf75290c9248e4d6bd76e480e7bdbfb8bc27e104297f44cfc1afc687eR1-R31) [[2]](diffhunk://#diff-7271320ec708563095a5ca533e73bba7707f8a09c5ef9ca692f2d0f6f1d36a2aR1-R83)
* Added the `CircularBuffer` class to `include/rtype/network/CircularBuffer.hpp` and implemented its methods in `src/network/CircularBuffer.cpp` for efficient circular byte storage and retrieval. [[1]](diffhunk://#diff-a9cb0a66048b55641840930283efabf3f769f52fc68469fc5e938cc9a72a965bR1-R38) [[2]](diffhunk://#diff-39e9af475166977c8e0e60fdbf6926610fedfa940ef4a4573e6856fad5af5938R1-R70)

### Integration

* Integrated `CircularBuffer` into the `UdpSocket` class by including its header and adding a `receiveBuffer` member, preparing for buffered network data handling. [[1]](diffhunk://#diff-adc9a26adb6668ad5d764f40cffecf39dc0069aff1b589ff753b265a68d6acdfR12) [[2]](diffhunk://#diff-adc9a26adb6668ad5d764f40cffecf39dc0069aff1b589ff753b265a68d6acdfR31)
* Updated the CMake configuration in `src/network/CMakeLists.txt` to include the new source files for compilation.

### Testing

* Added comprehensive unit tests in `tests/network/test_serialization.cpp` to verify the correctness of both `ByteConverter` and `CircularBuffer` functionalities, including edge cases and multiple value serialization.

### CODE REVIEW

File: src/network/ByteConverter.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    buffer->push_back((uvalue >> 8) & 0xFF);
+    buffer->push_back((uvalue >> 16) & 0xFF);
+    buffer->push_back((uvalue >> 24) & 0xFF);
+}
+
+void ByteConverter::serializeString(std::shared_ptr<std::vector<uint8_t>> buffer, const std::string& value) {
+    if (!buffer)
+        return;
+    uint32_t length = static_cast<uint32_t>(value.size());
+    serializeInt(buffer, static_cast<int32_t>(length));
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/ByteConverter.cpp]: Fixed

File: include/rtype/network/CircularBuffer.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    private:
+        size_t nextIndex(size_t index) const;
+
+        std::vector<uint8_t> _buffer;
+        size_t _head;
+        size_t _tail;
+        size_t _size;
+        size_t _capacity;
+};
+
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply include/rtype/network/CircularBuffer.hpp]: Nope this bloats the file so much that we can't tell if it's a .md or a .hpp

File: src/network/ByteConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    float value;
+    std::memcpy(&value, &uvalue, sizeof(float));
+    return value;
+}
+
+std::string ByteConverter::deserializeString(const std::vector<uint8_t>& buffer, std::shared_ptr<size_t> offset) {
+    if (!offset) {
+        throw std::invalid_argument("Offset shared_ptr is null");
+    }
+    uint32_t length = static_cast<uint32_t>(deserializeInt(buffer, offset));
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/ByteConverter.cpp]: Fixed

File: src/network/CircularBuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    for (uint8_t byte : data) {
+        if (_size == _capacity) {
+            _tail = nextIndex(_tail);
+        } else {
+            ++_size;
+        }
+        _buffer[_head] = byte;
+        _head = nextIndex(_head);
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/CircularBuffer.cpp]: This is a circular buffer it overwrites itself it's he's whole purpose it's a data saving mechanism

File: src/network/ByteConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+namespace rtype::network {
+
+void ByteConverter::serializeInt(std::shared_ptr<std::vector<uint8_t>> buffer, int32_t value) {
+    if (!buffer) return;
+    uint32_t uvalue = static_cast<uint32_t>(value);
+    buffer->push_back(uvalue & 0xFF);
+    buffer->push_back((uvalue >> 8) & 0xFF);
+    buffer->push_back((uvalue >> 16) & 0xFF);
+    buffer->push_back((uvalue >> 24) & 0xFF);
+}
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/ByteConverter.cpp]: That's a good comment this is fixed now we now use big endian as this is the primary network byte order for cross platform compatibility

File: src/network/CircularBuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+bool CircularBuffer::write(const std::vector<uint8_t>& data) {
+    for (uint8_t byte : data) {
+        if (_size == _capacity) {
+            _tail = nextIndex(_tail);
+        } else {
+            ++_size;
+        }
+        _buffer[_head] = byte;
+        _head = nextIndex(_head);
+    }
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/CircularBuffer.cpp]: This isn't an optimisation it's just trying to solve a problem that doesn't exist the principle of a circular buffer is to overwrite itself in a predictable manner so that isn't a correct correction you just gave me

File: src/network/ByteConverter.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                      static_cast<uint32_t>(buffer[*offset + 3]);
+    *offset += 4;
+    float value;
+    std::memcpy(&value, &uvalue, sizeof(float));
+    return value;
+}
+
+std::string ByteConverter::deserializeString(const std::vector<uint8_t>& buffer, std::shared_ptr<size_t> offset) {
+    if (!offset) {
+        throw std::invalid_argument("Offset shared_ptr is null");
```
[the remark this answers is not in the document]
Comment (ctalneau) [reply src/network/ByteConverter.cpp]: Too vague this is a school project also

**Review by SamTess — APPROVED** [review]
